pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -163,249 +352,157 @@ func (ap *ApplicationProfileCacheImpl) indexContainerCallStacks(containerID, con
 	for _, c := range appProfile.Spec.EphemeralContainers {
 		if c.Name == containerName {
 			for _, stack := range c.IdentifiedCallStacks {
-				index.searchTree.AddCallStack(stack)
+				callStackSearchTree.AddCallStack(stack)
 			}
+
+			// Clear the call stacks to free memory
 			c.IdentifiedCallStacks = nil
 			break
 		}
 	}
 }
 
-func (ap *ApplicationProfileCacheImpl) addApplicationProfile(obj runtime.Object) {
-	appProfile := obj.(*v1beta1.ApplicationProfile)
-	apName := objectcache.MetaUniqueName(appProfile)
-
-	if isUserManagedProfile(appProfile) {
-		ap.handleUserManagedProfile(appProfile)
-		return
-	}
-
-	// Original behavior for normal profiles
-	apState := newApplicationProfileState(appProfile)
-	ap.slugToState.Set(apName, apState)
-
-	if apState.status != helpersv1.Completed {
-		if ap.slugToAppProfile.Has(apName) {
-			ap.slugToAppProfile.Delete(apName)
-			ap.allProfiles.Remove(apName)
-		}
-		return
+// ContainerCallback handles container lifecycle events
+func (apc *ApplicationProfileCacheImpl) ContainerCallback(notif containercollection.PubSubEvent) {
+	switch notif.Type {
+	case containercollection.EventTypeAddContainer:
+		go func() {
```

Comment:
now that we have the async calls, I don't think you need the goroutine here

Reply:
You are probably right, but, I think it's a good habit to not have blocking things in the ContainerCallback.

Reply:
Removed.